verified_boot_app.md

```diff
@@ -57,12 +173,112 @@ Src: https://github.com/Kicksecure/kicksecure-wiki-backup/blob/6900fdfbdea8693fd
+     allowing to bypass signature verification when UEFI Secure Boot is
+     enabled.
+
+## Usecase examples
+
+<!--
 * Technical high-level examples:
 ** Power on -> firmware detects Microsoft Windows (bootloader) is installed -> show the key to the user -> user confirms or denies the key -> continue boot (if confirmed, otherwise go back to firmware menu or something)
 ** Power on -> firmware detects a "standard" Linux distribution such as Debian without a key in the well-known location -> show default Microsoft signing key for shim to the user -> user confirms or denies the key -> continue boot (if confirmed, otherwise go back to firmware menu or something)
 ** Power on -> firmware detects an "augmented" Linux distribution such as Kicksecure with a key database in the well-known location -> show key from key database to the user -> user confirms or denies the key -> continue boot (if confirmed, otherwise go back to firmware menu or something)
 ** User disables Secure Boot -> as usual
-* Additional notes:
-** To handle the situation where multiple shims are installed, the firmware may need the ability to find multiple candidate bootloaders in the removable media path (or any bootloader path)
+-->
 
 -->
+
+### Initial configuration
+
+<!-- TODO: diagrams? -->
+
+1. User powers on the platform.
+2. Firmware performs boot media enumeration.
+3. Firmware detects initial boot or boot with default settings and launches an
+   application.
+4. The application visits all handles with `gEfiSimpleFileSystemProtocolGuid`
+   and `gEfiPartTypeSystemPartGuid`.
+5. For each handle, find operating system bootloader. E.g. GRUB, SHIM, Windows
+   Boot Manager.
+6. For each handle, search for files containing keys and certificates. E.g.
+   files with `.der`, `.cer`, `.key` extensions.
+7. For each bootloader found in step 5, match a key found in step 6 or
+   `PKDefault`, `KEKDefault`, `dbDefault`, `dbxDefault`.
+8. If key not in `dbxDefault`, attempt image verification and report the
+   result to the user.
+9. If verification passed, ask the user to trust the key and optionally offer
+   to add a boot option.
+11. If the user decides to trust the key, the key is added to `DB`.
+12. Ask the user to continue boot or proceed with next bootloader verification
+    (go to step 8).
+13. If the user decided to continue boot, ask the user to enroll a `PK` from a
+    pool of detected keys. Then enable Secure Boot and proceed with booting.
+14. If no more bootloaders present and user did not trust any key, go to the
+    firmware setup menu.
+
+### Secure Boot fails to verify the image
+
+1. User powers on the platform.
+2. Firmware performs boot media enumeration.
+3. Firmware attempts to boot a boot option according to boot order.
+4. Image pointed by the boot option fails the verification.
+5. The firmware launches the application to handle possibly new operating
+   system or a change in the operating system keys/components.
```

Comment:
How would firmware know that? I think there is edk2 way to handle recovery, but I cannot recall the correct flow. I have one slide in DS08MSA.

Reply:
It is described in https://github.com/3mdeb/verified-boot/pull/31/files#diff-38c2b0b3bb94a635d00a3454d8f6a30853d3036a45fb5d64317ff160771ef2d4R200

Reply:
Ok, didn't get there yet. Maybe we should decouple to not confuse, or maybe we should start with Integration Consideration and reverse the spec?

Reply:
Maybe the whole suecase section should be dropped from here and just be a part of the spec...